File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    const len = triangle.length;
    const dp = Array.from({length: len}, ()=> 0)
    
    dp[0] = triangle[0][0];
    
    for(let i=1; i<len; i++){
        for(let j=i; j>=0; j--){
            if(j===0)   dp[j] = triangle[i][j] + dp[0];
            else if(j===i)  dp[j] = triangle[i][j] + dp[j-1];
            else    dp[j] = triangle[i][j] + Math.max(dp[j-1], dp[j]);
        }
    }
        
    for(let i=0; i<dp.length; i++){
        if(answer<dp[i])    answer = dp[i];
    }
    
    
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    N = len(triangle)
    dp = [[0] * N for _ in range(N)]
    dp[0][0] = triangle[0][0]

    
    for i in range(1, N):
        for j in range(len(triangle[i])):
            if j==0:
                dp[i][0] = dp[i-1][0] + triangle[i][j]
            elif i==j:
                dp[i][j] = dp[i-1][i-1] + triangle[i][j]
            else:
                dp[i][j] = max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j]
    
    return max(dp[N-1][:])